File: matrix.py
import numpy as np
from scipy import sparse
import time
import os
from multiprocessing import Pool, cpu_count, Lock, Value

sizes = [64, 128, 256, 512, 768, 1024]
runs = 5
sparsity_levels = [0.1, 0.5, 0.9]

#Paralellization
def dense_row(args):
    A, B, i = args
    return np.dot(A[i], B)

def parallel_dense(A, B):
    N = A.shape[0]
    with Pool(cpu_count()) as p:
        rows = p.map(dense_row, [(A, B, i) for i in range(N)])
    return np.vstack(rows)

def sparse_row(args):
    A, B, i = args
    return A[i].dot(B)

def parallel_sparse(A, B):
    N = A.shape[0]
    with Pool(cpu_count()) as p:
        rows = p.map(sparse_row, [(A, B, i) for i in range(N)])
    return sparse.vstack(rows)

def parallel_optimized(A, B):
    return A @ B

lock = Lock()
counter = Value('i', 0)

def increment_counter(_):
    with lock:
        counter.value += 1


if __name__ == "__main__":

    os.makedirs("results", exist_ok=True)

    with open("results/python_parallel.txt", "w") as f:
        #Dense
        for n in sizes:
            times = []
            f.write(f"\n=== Dense {n}x{n} ===\n")
            print(f"\n=== Dense {n}x{n} ===")

            for r in range(runs):
                A = np.random.rand(n, n)
                B = np.random.rand(n, n)

                start = time.time()
                parallel_dense(A, B)
                end = time.time()

                elapsed = round(end - start, 3)
                times.append(elapsed)

                f.write(f"Run {r+1}: {elapsed} s\n")
                print(f"Run {r+1}: {elapsed} s")

            f.write(f"Mean: {round(sum(times)/runs,3)} s\n")
        #Sparsity
        for sp in sparsity_levels:
            f.write(f"\n=== Sparse (sparsity = {sp}) ===\n")
            print(f"\n=== Sparse (sparsity = {sp}) ===")

            for n in sizes:
                f.write(f"--- N={n} ---\n")
                times = []

                for r in range(runs):
                    density = 1 - sp
                    A = sparse.random(n, n, density=density, format='csr')
                    B = sparse.random(n, n, density=density, format='csr')

                    start = time.time()
                    parallel_sparse(A, B)
                    end = time.time()

                    elapsed = round(end - start, 3)
                    times.append(elapsed)

                    f.write(f"Run {r+1}: {elapsed} s\n")
                    print(f"Run {r+1}: {elapsed} s")

                f.write(f"Mean: {round(sum(times)/runs,3)} s\n")
        #Optimzed
        for n in sizes:
            times = []
            f.write(f"\n=== Optimized {n}x{n} ===\n")
            print(f"\n=== Optimized {n}x{n} ===")

            for r in range(runs):
                A = np.random.rand(n, n)
                B = np.random.rand(n, n)

                start = time.time()
                parallel_optimized(A, B)
                end = time.time()

                elapsed = round(end - start, 3)
                times.append(elapsed)

                f.write(f"Run {r+1}: {elapsed} s\n")
                print(f"Run {r+1}: {elapsed} s")

            f.write(f"Mean: {round(sum(times)/runs,3)} s\n")

    with Pool(cpu_count()) as p:
        p.map(increment_counter, range(10000))

    print("\nSynchronized counter:", counter.value)

File: Matrix.java
import java.io.FileWriter;
import java.io.IOException;
import java.util.stream.IntStream;

public class Matrix {

    static int RUNS = 5;
    static double[][] randomDense(int N) {
        double[][] M = new double[N][N];
        for(int i=0;i<N;i++)
            for(int j=0;j<N;j++)
                M[i][j] = Math.random();
        return M;
    }

    static double[][] randomSparse(int N, double sparsity) {
        double[][] M = new double[N][N];
        double density = 1.0 - sparsity;

        for(int i=0;i<N;i++)
            for(int j=0;j<N;j++)
                M[i][j] = (Math.random() < density) ? Math.random() : 0.0;

        return M;
    }

    static double[][] multiplyDenseParallel(double[][] A, double[][] B) {
        int N = A.length;
        double[][] C = new double[N][N];

        IntStream.range(0, N).parallel().forEach(i -> {
            for (int j = 0; j < N; j++) {
                double sum = 0;
                for (int k = 0; k < N; k++)
                    sum += A[i][k] * B[k][j];
                C[i][j] = sum;
            }
        });
        return C;
    }

    static double[][] multiplySparse(double[][] A, double[][] B) {
        int N = A.length;
        double[][] C = new double[N][N];

        for(int i=0;i<N;i++) {
            for(int j=0;j<N;j++) {
                double sum = 0;
                for(int k=0;k<N;k++) {
                    if(A[i][k] != 0 && B[k][j] != 0)
                        sum += A[i][k] * B[k][j];
                }
                C[i][j] = sum;
            }
        }
        return C;
    }

    static double[][] multiplyOptimized(double[][] A, double[][] B) {
        int N = A.length;
        double[][] Bt = new double[N][N];

        // transpose B
        for(int i=0;i<N;i++)
            for(int j=0;j<N;j++)
                Bt[i][j] = B[j][i];

        double[][] C = new double[N][N];

        for(int i=0;i<N;i++)
            for(int j=0;j<N;j++){
                double sum=0;
                for(int k=0;k<N;k++)
                    sum += A[i][k] * Bt[j][k];
                C[i][j]=sum;
            }

        return C;
    }

    public static void main(String[] args) throws IOException {

        int[] sizes = {64,128,256,512,768,1024};
        double[] sparsityLevels = {0.1,0.5,0.9};

        new java.io.File("results").mkdirs();
        FileWriter fw = new FileWriter("results/java_parallel.txt");

        for(int N : sizes) {
            fw.write("\n=== Dense " + N + "x" + N + " ===\n");

            double total=0;
            for(int r=1;r<=RUNS;r++){
                double[][] A = randomDense(N);
                double[][] B = randomDense(N);

                long start = System.nanoTime();
                multiplyDenseParallel(A,B);
                long end = System.nanoTime();

                double sec = (end-start)/1e9;
                total += sec;

                fw.write("Run " + r + ": " + sec + " s\n");
            }
            fw.write("Mean: " + (total/RUNS) + " s\n");
        }

        for(double sp : sparsityLevels) {

            fw.write("\n=== Sparse (sparsity = " + sp + ") ===\n");

            for(int N : sizes) {

                fw.write("--- N=" + N + " ---\n");

                double total=0;

                for(int r=1;r<=RUNS;r++){
                    double[][] A = randomSparse(N, sp);
                    double[][] B = randomSparse(N, sp);

                    long start = System.nanoTime();
                    multiplySparse(A,B);
                    long end = System.nanoTime();

                    double sec = (end-start)/1e9;
                    total += sec;

                    fw.write("Run " + r + ": " + sec + " s\n");
                }

                fw.write("Mean: " + (total/RUNS) + " s\n");
            }
        }

        for(int N : sizes) {

            fw.write("\n=== Optimized " + N + "x" + N + " ===\n");

            double total=0;

            for(int r=1;r<=RUNS;r++){
                double[][] A = randomDense(N);
                double[][] B = randomDense(N);

                long start = System.nanoTime();
                multiplyOptimized(A,B);
                long end = System.nanoTime();

                double sec = (end-start)/1e9;
                total += sec;

                fw.write("Run " + r + ": " + sec + " s\n");
            }

            fw.write("Mean: " + (total/RUNS) + " s\n");
        }

        fw.close();
        System.out.println("Java results generated in results/java_parallel.txt");
    }
}